spinalcordtoolbox/deepseg_/gm.py

```diff
@@ -275,5 +276,6 @@ def segment_file(input_filename, output_filename,
     if threshold is not None:
         res_data = threshold_predictions(res_data, 0.5)
 
-    nib.save(nii_resampled_original, output_filename)
+    img = Image(param=res_data, hdr=nii_resampled_original.header)
+    img.save(output_filename)
```

Comment:
Wait, is this actually equivalent? In the old code, it looks like `res_data` is just thrown away after being computed? Or does it get mutated in some way that survives in `nii_resampled_original`? But that seems unlikely after getting passed through `np.asanyarray()`.

EDIT: Ah! looking at `help(np.asanyarray)`, it confirms that some cases get passed through as-is (so that mutations flow upstream), and some cases get turned into copies (so that mutations **don't** flow upstream). I'm not sure which case `.dataobj` falls under, or whether Nibabel makes any promises. Seems dangerous!

Reply:
Just to compare the before and after with the full context (since GitHub is doing a poor job showing the full diff):

```diff
nii_resampled_original = resampling.resample_nib(nii_segmentation, new_size=original_res, new_size_type='mm', interpolation='linear')
res_data = np.asanyarray(nii_resampled_original.dataobj)

# Threshold after resampling, only if specified
if threshold is not None:
    res_data = threshold_predictions(res_data, 0.5)

-nib.save(nii_resampled_original, output_filename)
+img = Image(param=res_data, hdr=nii_resampled_original.header)
+img.save(output_filename)
```

You're absolutely correct that `res_data` was thrown away previously. Though, I think there are actually 2 big differences that this change runs into:

1. The use of `np.asanyarray(img.dataobj)` (as you pointed out).
2. The call to `threshold_predictions`, which was completely ignored previously. (!!!)

To address the first point re: `np.asanyarray`, from what I understand, this is actually `nibabel`'s recommended method for extracting the data array from `Nifti1Image` objects. I think the reason I recognize this pattern is because of nibabel's previous deprecation of `get_data` and switch to `get_fdata` (https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3854). In that issue, we [came to the conclusion](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3854#issuecomment-1313976579) that `np.asanyarray` was the way to go to preserve the old `get_data` functionality. (And, just to clarify why we use `np.asanyarray` rather than just using `img.dataobj` directly -- I think this is because of Nibabel's memory-mapped array behavior?)

As for the second point -- I'm realizing this is a behavior change! (Though, arguably the fact that we weren't using the thresholded `res_data` smells like a bug to me? So, maybe this is a welcome change...)

Reply:
Ah! It's not a behavior change. I just checked, and all the operations here, including the ones which _look_ like they make copies, actually affect the original array in-place:
```py
>>> import numpy as np
>>> x = np.arange(5); x
array([0, 1, 2, 3, 4])
>>> y = np.asanyarray(x); y
array([0, 1, 2, 3, 4])
>>> z = y[:]; z
array([0, 1, 2, 3, 4])
>>> z[0] = 99
>>> z
array([99,  1,  2,  3,  4])
>>> y
array([99,  1,  2,  3,  4])
>>> x
array([99,  1,  2,  3,  4])
```
So, the original code _did_ threshold, although maybe it only succeeded accidentally.

Reply:
Huzzah! Thank you for checking. :heart: